main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Documents/DocumentModel.cs

```diff
@@ -0,0 +1,496 @@
+		{
+			CheckInitialized ();
+			await ModelRepresentation.Save ();
+		}
+
+		public async Task Reload ()
+		{
+			if (!Data.IsLinked)
+				throw new InvalidOperationException ("The model is not linked");
+			var rep = await GetRepresentationAsync ();
+			await rep.Reload ();
+		}
+
+		public async Task Load ()
+		{
+			CheckInitialized ();
+			modelRepresentation = await GetRepresentationAsync ();
+			await modelRepresentation.Load ();
+		}
+
+		public bool IsLoaded => modelRepresentation != null && modelRepresentation.IsLoaded;
+
+		public void Dispose ()
+		{
+			if (disposed)
+				return;
+			disposed = true;
+			if (Registry != null)
+				Registry.DisposeModel (this).Ignore ();
+			else
+				Data.Unlink (this).Ignore ();
+		}
+
+		public async Task Synchronize ()
+		{
+			CheckInitialized ();
+			var rep = await GetRepresentationAsync ();
+			await rep.Synchronize ();
+		}
+
+		protected Task Relink (object id)
+		{
+			isNew = false;
+			if (Registry != null)
+				return Registry.RelinkModel (this, id);
+			else {
+				Data.Id = id;
+				return Task.CompletedTask;
+			}
+		}
+
+		protected Task UnlinkFromId ()
+		{
+			isNew = true;
+			if (Registry != null)
+				return Registry.RelinkModel (this, null);
+			else {
+				Data.Id = null;
+				return Task.CompletedTask;
+			}
+		}
+
+		internal DocumentModelRegistry Registry { get; set; }
+
+		internal protected abstract Type RepresentationType { get; }
+
+
+		protected T GetRepresentation<T> () where T : ModelRepresentation
+		{
+			return (T)ModelRepresentation;
+		}
+
+		async Task<T> GetRepresentationAsync<T> () where T : ModelRepresentation
+		{
+			return (T) await Data.GetRepresentation (RepresentationType);
+		}
+
+		Task<ModelRepresentation> GetRepresentationAsync ()
+		{
+			return Data.GetRepresentation (RepresentationType);
+		}
+
+		internal protected virtual void OnRepresentationChanged ()
+		{
+		}
+
+		internal async Task RaiseRepresentationChangeEvent ()
+		{
+			if (modelRepresentation == null)
+				return;
+
+			var newRep = await GetRepresentationAsync ();
+			if (newRep != modelRepresentation) {
+				modelRepresentation = newRep;
+				try {
+					OnRepresentationChanged ();
+				} catch (Exception ex) {
+					LoggingService.LogError ("OnRepresentationChanged failed", ex);
+				}
+				RaiseChangeEvent ();
+			}
+		}
+
+		internal void RaiseChangeEvent ()
+		{
+			if (modelRepresentation != null && notifiedChangeVersion != modelRepresentation.CurrentVersion) {
+				try {
+					notifiedChangeVersion = modelRepresentation.CurrentVersion;
+					Changed?.Invoke (this, EventArgs.Empty);
+				} catch (Exception ex) {
+					LoggingService.LogError ("RaiseChanged failed", ex);
+				}
+			}
+		}
+
+		internal class DocumentModelData
+		{
+			ImmutableList<DocumentModel> linkedModels = ImmutableList<DocumentModel>.Empty;
+			Dictionary <Type, ModelRepresentation> representations = new Dictionary<Type, ModelRepresentation> ();
+			SemaphoreSlim dataLock = new SemaphoreSlim (1);
```

Comment:
Unless we set the maximum count too, releasing dataLock twice in a row will allow 2 concurrent accesses.

```suggestion
			SemaphoreSlim dataLock = new SemaphoreSlim (1, 1);
```